readme.md
# PostHTML Picture Srcset <img align="right" height="100" title="PostHTML logo" src="http://posthtml.github.io/posthtml/logo.svg">

[![NPM][npm]][npm-url]

This postHTML plugin uses rollup-images to generate webp versions of your png and jpeg images in multiple responsive sizes. It then replaces the reference to that image in your HTML with the corresponding `<picture>` tag and the corresponding `srcset` attributes.

Also includes fallback to the original image for older browsers.

Before:
``` html
<img src="bigimg.jpg" />
```

After:
``` html
<picture>
    <source data-srcset="/bigimg@640w.webp 640w, /bigimg@1300w.webp 1300w, /bigimg@1800w.webp 1800w, /bigimg.webp 1900w" type="image/webp" srcset="/bigimg@640w.webp 640w, /bigimg@1300w.webp 1300w, /bigimg@1800w.webp 1800w, /bigimg.webp 1900w">
    <source data-srcset="/bigimg.png" type="image/png" srcset="/bigimg.png">
    <img data-src="/bigimg.webp" skip width="1900" height="1200" nolazy="" alt="" src="/bigimg.webp">
</picture>
```

## Install

```bash
npm i posthtml-picture-srcset
```

## Usage

Describe how people can use this plugin. Include info about build systems if it's
necessary.

``` js
const fs = require('fs');
const posthtml = require('posthtml');
const pictureSrcset = require('posthtml-picture-srcset');

posthtml()
    .use(
      pictureSrcset({
        cwd: resolve('src/assets/images'),
        imgDir: resolve('src/assets/images'),
        imageOptions: {
          generate: true,
          dir: ['src/assets/images'],
          size: [640, 1300, 1800],
          hook: 'renderStart',
          quality: 80,
          inputFormat: ['jpg', 'jpeg', 'png'],
          outputFormat: ['webp'],
          maxParallel: 25,
          forceUpscale: false,
          skipExisting: false,
        },
    )
    .process(html/*, options */)
    .then(result => fs.writeFileSync('./after.html', result.html));


```

## Global Options

| Option      | Type | Description  |
| ----------- | ----------- |  ----------- |
| cwd         | path        | The current working directory |
| imgDir      | path        | Image folder  |
| imageOptions| object      | See image options below |

## Image Options
| Option      | Type | Description  |
| ----------- | ----------- |  ----------- |
| generate    | path        | The current working directory |
| dir         | array       | Pass in the image folder  |
| size        | array       | Array of image sizes to generate|
| hook        | string      | Rollup hook, when to generate images  |
| quality     | number      | Value between 1 and 100 for image generation quality  |
| inputFormat | array       | Image formats to run on  |
| outputFormat| array       | Image formats to generate |
| maxParallel | number      | Maximum number of images to generate on one thread  |
| forceUpscale| boolean     | Generate larger images if a file resolution is less than the largest output size  |
| skipExisting| boolean     | Skip existing images  |



### Contributing

See [PostHTML Guidelines](https://github.com/posthtml/posthtml/tree/master/docs) and [contribution guide](CONTRIBUTING.md).

[action]: https://github.com/kingkongdevs/posthtml-picture-srcset/workflows/Actions%20Status/badge.svg
[action-url]: https://github.com/kingkongdevs/posthtml-picture-srcset/actions?query=workflow%3A%22CI+tests%22

[npm]: https://img.shields.io/npm/v/posthtml-picture-srcset.svg
[npm-url]: https://npmjs.com/package/posthtml-picture-srcset


changelog.md
## <small>1.0.2 (2023-07-24)</small>

* 🎉 initial commit ([b51ec0e](https://github.com/kingkongdevs/posthtml-picture-srcset/commit/b51ec0e))
* 🐛 fixed the export reference ([7419fbb](https://github.com/kingkongdevs/posthtml-picture-srcset/commit/7419fbb))
* 📝 added repo link ([5f75b2a](https://github.com/kingkongdevs/posthtml-picture-srcset/commit/5f75b2a))
* 📝 updated documentation ([c44e92f](https://github.com/kingkongdevs/posthtml-picture-srcset/commit/c44e92f))
* 📝 updated package description and name ([09aeec0](https://github.com/kingkongdevs/posthtml-picture-srcset/commit/09aeec0))
* 🔧 added type of module ([11c3c22](https://github.com/kingkongdevs/posthtml-picture-srcset/commit/11c3c22))





lib/picture-srcset.js
const path = require('path')
const fs = require('fs')
const parser = require('posthtml-parser')
const match = require('posthtml/lib/api').match
const sizeOf = require('image-size')
const generateAll = require('./inc/rollup-images.js').generateAll;

module.exports = function (options) {
  const cwd = options.cwd || process.cwd()
  const imageOptions = options.imageOptions || {}

  if (!options) {
    options = {}
  }

  if (imageOptions.size === undefined) {
    imageOptions.size = [320, 640, 1024];
  }

  if (options.imgDir === undefined) {
    options.imgDir = './dist/assets'
  }

  if (options.format === undefined) {
    options.format = '.webp'
  }

  const getAssets = (imageDir) => {
    return fs.readdirSync(imageDir) // Read the contents of the assets folder
  }

  const getPicture = async (img, src) => {

    // If the image is referencing an external url, skip it
    if (img.attrs.src.startsWith('http://') || img.attrs.src.startsWith('https://')) {
      return img
    }

    // If the image has a skip flag, skip it
    if(img.attrs['skip'] == true) {
      return img
    }

    // Loop through the different set sizes and check if an image exists for that size
    let srcset = []
    var dimensions = sizeOf(src)
    var relativeSrc = src.split('src')[1]
    img.content = [];

    if (src.indexOf('.gif') != -1) {
      img['attrs']['src'] = relativeSrc
      return img
    }

    if(typeof(img['attrs']['fullsize']) == "undefined"){
      // Loop through the assets folder and see if each size exists in webp format
      for (var i = 0; i < imageOptions.size.length; i++) {
        var size = imageOptions.size[i]
        var suffix =  '@' + size + 'w';
        var image =
          relativeSrc.replace(/\.(\w+)$/, suffix + options.format) +
          ' ' +
          size +
          'w'

        if(!(size > dimensions.width)) {
          srcset.push(image)
        }
      }
    }

    // add full size to images array
    srcset.push(relativeSrc.replace(/\.(\w+)$/, options.format+' '+dimensions.width+'w'));

    // Generate the picture tag with the <source> elements + the <img> element
    img.tag = 'picture'
    delete img['attrs']['src'];

    // Add an alt tag if it's left empty
    if(img['attrs']['alt'] === undefined) {
      img['attrs']['alt'] = ''
    }


    if(img['attrs']['nolazy'] !== undefined) {
      img['attrs']['src'] = relativeSrc.replace(/\.(\w+)$/, options.format)
    } else {
      if(typeof(img['attrs']['class']) == 'undefined'){
        img['attrs']['class'] = '';
      }

      img['attrs']['class'] += ' lazyload';
    }

    // Set the srcset <source> tags
    let srcsetTag = {
      tag: 'source',
      attrs: {
        "data-srcset": srcset.join(', '),
        type: 'image/webp'
      },
    };


    // Set the fallback image

    let fallbackType = relativeSrc.split('.')[1];
    if(fallbackType == 'jpg'){
      fallbackType = 'jpeg';
    }

    let defaultSrc =  {
      tag: 'source',
      attrs: {
        "data-srcset": relativeSrc,
        type: 'image/'+ fallbackType
      },
    };

    if(img['attrs']['nolazy'] !== undefined) {
      defaultSrc['attrs']['srcset'] = defaultSrc['attrs']['data-srcset'];
      srcsetTag['attrs']['srcset'] = srcsetTag['attrs']['data-srcset'];
    }

    img.content.push([
      srcsetTag,
      defaultSrc,
    ]);

    img.content.push([
      {
        tag: 'img',
        attrs: {
          "data-src": relativeSrc.replace(/\.(\w+)$/, '.webp'),
          skip: true,
          width: dimensions.width,
          height: dimensions.height,
          ...img['attrs']
        },
      },
    ]);

    if (imageOptions.generate == false) {
      img['attrs']['src'] = relativeSrc
      img.content = [
        {
          tag: 'img',
          attrs: {
            "data-src": relativeSrc,
            skip: true,
            width: dimensions.width,
            height: dimensions.height,
            ...img['attrs']
          },
        },
      ]
    }

    delete img['attrs'];
    return img
  }

  /**
   * Ease of use function to generate all the images + sizes
   */
  const generateImages = async () => {
    if (imageOptions.generate) {
      return await generateAll(imageOptions)
    }
  }

  return (tree) =>
    new Promise((resolve, reject) => {
      // generates the required image files
      generateImages()

      const promises = []
      if (!tree.promises) tree.parser = parser
      if (!tree.match) tree.match = match

      tree.match({ tag: 'img' }, (node) => {
        if (!node.attrs['skip'] && !node.attrs.src?.includes('.svg')) {
          promises.push(
            new Promise(async (resolve, reject) => {
              try {
                var src = node.attrs['src']
                if (src) {
                  var imgSrc = path.resolve(cwd, src)
                  // Get the picture tag and assign it
                  node = await getPicture(node, imgSrc)
                }
                resolve()
              } catch (err) {
                reject(err)
              }
            })
          )
        }
        return node
      })
      Promise.all(promises)
        .then(() => resolve(tree))
        .catch(reject)
    })
}


lib/inc/rollup-images.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var fs = require('fs');
var fse = require('fs-extra');
const path = require('path');
const sizeOf = require('image-size')
var globby = require('globby');
const sharp = require('sharp');
var Queue = require('promise-queue');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var fs__default = /*#__PURE__*/_interopDefaultLegacy(fs);
var globby__default = /*#__PURE__*/_interopDefaultLegacy(globby);
// var jimp__default = /*#__PURE__*/_interopDefaultLegacy(jimp);
var Queue__default = /*#__PURE__*/_interopDefaultLegacy(Queue);

/**
 * A helper function for transform single items to array such that:
 * [a] => [a]
 * a => [a]
 * ...
 *
 * @param a
 */
const arrayify = (a) => (Array.isArray(a) ? [...a] : [a]);

/**
 * Generates the crops for this image and saves it to disk.
 *
 * @param image
 * @param outputs
 * @param quality
 * @param forceUpscale
 * @param imagePathPre
 * @param addGeneratedFile
 */
const generateCrops = ({
  image, outputs, imagePathPre, quality, forceUpscale, addGeneratedFile,
}) => () => sharp(image).webp({ lossless: false }).toFile(image.replace(/(\..*)/, '.webp'))
  .then((webpImage) => Promise.allSettled(
    (outputs)

      // Get only the sizes that we need to generate
      .reduce((acc, val) => {
        if (acc.indexOf(val.scaleWidth) < 0) return [...acc, val.scaleWidth];
        return acc;
      }, [])
      .map((scaleWidth, i) => {

        var outputWebp = image.replace(/(\..*)/, '.webp');
        var imageDimensions = {'width':0};

        // todo - there's a missing file error here
        try{
          imageDimensions = sizeOf(outputWebp);
        } catch(e){}

        // If the target width is larger than the original
        // width and forceUpscale is not set, we skip this.
        var imagePathPost = imagePathPre.replace('src', 'dist');

        // if ((i == 0)) {
        //   let j = 0;
        //   outputs.forEach((d, i) => {
        //     if(d.format == 'webp') {
        //       jimpObj.clone().write(`${imagePathPost}.${d.format}`);
        //       j++;
        //     }
        //   })
        // }

        if(scaleWidth > imageDimensions.width) {
          return Promise.resolve();
        }

        // Save all the output images
        return Promise.all(
          outputs
            // only get the outputs of the current width
            .filter((d) => d.scaleWidth === scaleWidth)
            .map((d) => d.format)
            .map((format) => {

                // Add this file to the generated file list
                addGeneratedFile(image , scaleWidth, format);

                return sharp(outputWebp)
                  .resize(scaleWidth)
                  .toFile(`${imagePathPost}@${scaleWidth}w.${format}`);

                // // Do the resize and save
                // return jimp__default["default"]
                //   .clone()
                //   .resize(scaleWidth, jimp__default["default"].AUTO)
                //   .write(`${imagePathPost}@${scaleWidth}w.${format}`);
            }),
        );
      }),
  ));

/**
 * Processes this image, determining which crops and sizes we need to generate.
 *
 * @param queue
 * @param size
 * @param skipExisting
 * @param outputFormat
 * @param quality
 * @param forceUpscale
 * @param addGeneratedFile
 * @returns {(function(*, *, *): void)|*}
 */
const processImage = ({
  queue, size, skipExisting, outputFormat, quality, forceUpscale, addGeneratedFile,
}) => (image, index, imageArray) => {
  const sizes = arrayify(size);

  // generate the output path
  const imagePathSplit = image.split('.');
  const imagePathPre = imagePathSplit.slice(0, -1).join('.');
  const imageFormat = imagePathSplit[imagePathSplit.length - 1];

  // process image format options
  const formats = Array.from(new Set(
    arrayify(outputFormat)
      // If format is match, we match to the input format
      .map((format) => (format === 'match' ? imageFormat : format))
      // If format is jpeg, we map to jpg
      .map((format) => (format === 'jpeg' ? 'jpg' : format)),
  ));

  // An array of objects that contains sizes and formats of all our outputs.
  let outputs = sizes.reduce(
    (acc, scaleWidth) => [
      ...acc,
      ...formats.map((format) => ({ format, scaleWidth })),
    ],
    [],
  );

  // if skipExisting is set
  if (skipExisting) {
    // Filter out images that already exist
    outputs = outputs.filter(
      ({ format, scaleWidth }) => {
        const cropExists = fs__default["default"].existsSync(`${imagePathPre}@${scaleWidth}w.${format}`);
        // If this image crop already exists, we add it to the generated image list anyway
        if (cropExists) addGeneratedFile(image, scaleWidth, format);
        return !cropExists;
      },
    );

    // if images already exist, we can skip this rest of this process
    if (outputs.length === 0) {
      // If every single file is skipped, we still need to resolve
      // so that promise-queue doesn't sit around waiting forever.
      if (index === imageArray.length - 1) queue.add(() => Promise.resolve());
      return;
    }
  }

  // copy the raw unprocessed image to the dist folder
  fse.copySync(image, image.replace('src', 'dist'));

  // update current image url to the dist version
  image = image.replace('src', 'dist');

  // load in the image
  queue.add(generateCrops({
    image, outputs, imagePathPre, quality, forceUpscale, addGeneratedFile,
  }));

};

const generateAll = (options) => {
  // Load options
  const {
    hook = 'renderStart', // rollup hook
    quality = 65, // image quality
    dir = null, // directory string or strings,
    src = null, // directory string or strings,
    size = [1400, 1024, 640, 320], // image size or sizes
    inputFormat = ['jpg', 'jpeg', 'png'], // image input formats
    outputFormat = ['jpg'], // image output formats
    forceUpscale = false, // whether we should forcibly upscale
    skipExisting = true, // whether we should skip existing images that have already been resized
    maxParallel = 4, // the max number of parallel images that can be processed concurrently
    outputManifest = null, // The path to output a manifest of all generated images
  } = options;

  // If preconditions are not met, we just quit.
  if (
    !dir || dir.length === 0
    || !size
    || !inputFormat || inputFormat.length === 0
    || !outputFormat || outputFormat.length === 0
  ) { return Promise.resolve(); }

  // The sizes and formats that we will output to, arrayified.
  const inputFormats = arrayify(inputFormat);

  // Glob for the directories
  const dirGlob = arrayify(dir).length > 1 ? `{${arrayify(dir).join(',')}}` : arrayify(dir)[0];

  // An array that contains all the images we've generated (or already exist)
  let generatedFiles = [];
  const addGeneratedFile = (image, width, format) => {
    generatedFiles = [...generatedFiles, { image, width, format }];
  };

  return new Promise((resolve) => {
    // Promise queue so that we only process maxParallel parallel files at a time
    const q = new Queue__default["default"](maxParallel, Infinity, {
      onEmpty: () => {
        if (outputManifest) fs__default["default"].writeFileSync(outputManifest, JSON.stringify(generatedFiles));
        return resolve();
      },
    });

    // Finds all the images we want based on dir and inputFormat
    globby__default["default"](`${dirGlob}/**/*.{${inputFormats.join(',')}}`)
      .then((images) => {
        // If no images were provided, resolve the queue.
        if (images.length === 0) q.add(() => Promise.resolve());

        // Map them into jimp objects
        images
          .filter((d) => d.indexOf('@') < 0 && d.indexOf('#') < 0)
          .forEach(processImage({
            queue: q, size, skipExisting, outputFormat, quality, forceUpscale, addGeneratedFile,
          }));
      });
  });
}

/**
 * Generate different image sizes on all images found within a directory,
 * based on the configuration given. Uses jimp as the image processing module.
 *
 * @param options
 * @returns {Promise<unknown>|Promise<void>|{name: string}}
 */
const generateImageSizes = (options) => {
  // Load options
  const {
    hook = 'renderStart', // rollup hook
    quality = 65, // image quality
    dir = null, // directory string or strings
    size = [1400, 1024, 640, 320], // image size or sizes
    inputFormat = ['jpg', 'jpeg', 'png'], // image input formats
    outputFormat = ['jpg'], // image output formats
    forceUpscale = false, // whether we should forcibly upscale
    skipExisting = true, // whether we should skip existing images that have already been resized
    maxParallel = 4, // the max number of parallel images that can be processed concurrently
    outputManifest = null, // The path to output a manifest of all generated images
  } = options;

  return {
    name: 'generate-image-sizes',

    // Runs on the hook specified, otherwise the renderStart hook.
    // First, get all the image files in the given directory
    [hook]() {
      // If preconditions are not met, we just quit.
      if (
        !dir || dir.length === 0
        || !size
        || !inputFormat || inputFormat.length === 0
        || !outputFormat || outputFormat.length === 0
      ) { return Promise.resolve(); }

      // The sizes and formats that we will output to, arrayified.
      const inputFormats = arrayify(inputFormat);

      // Glob for the directories
      const dirGlob = arrayify(dir).length > 1 ? `{${arrayify(dir).join(',')}}` : arrayify(dir)[0];

      // An array that contains all the images we've generated (or already exist)
      let generatedFiles = [];
      const addGeneratedFile = (image, width, format) => {
        generatedFiles = [...generatedFiles, { image, width, format }];
      };

      return new Promise((resolve) => {
        // Promise queue so that we only process maxParallel parallel files at a time
        const q = new Queue__default["default"](maxParallel, Infinity, {
          onEmpty: () => {
            if (outputManifest) fs__default["default"].writeFileSync(outputManifest, JSON.stringify(generatedFiles));
            return resolve();
          },
        });

        // Finds all the images we want based on dir and inputFormat
        globby__default["default"](`${dirGlob}/**/*.{${inputFormats.join(',')}}`)
          .then((images) => {
            // If no images were provided, resolve the queue.
            if (images.length === 0) q.add(() => Promise.resolve());

            // Map them into jimp objects
            images
              .filter((d) => d.indexOf('@') < 0 && d.indexOf('#') < 0)
              .forEach(processImage({
                queue: q, size, skipExisting, outputFormat, quality, forceUpscale, addGeneratedFile,
              }));
          });
      });
    },
  };
};

exports.generateAll = generateAll;
exports.arrayify = arrayify;
exports.generateCrops = generateCrops;
exports.generateImageSizes = generateImageSizes;
exports.processImage = processImage;


lib/pictureSrcset.js
const path = require('path')
const fs = require('fs')
const parser = require('posthtml-parser')
const match = require('posthtml/lib/api').match
const sizeOf = require('image-size')
const generateAll = require('./inc/rollup-images.js').generateAll;

module.exports = function (options) {
  const cwd = options.cwd || process.cwd()
  const imageOptions = options.imageOptions || {}

  if (!options) {
    options = {}
  }

  if (imageOptions.size === undefined) {
    imageOptions.size = [320, 640, 1024];
  }

  if (options.imgDir === undefined) {
    options.imgDir = './dist/assets'
  }

  if (options.format === undefined) {
    options.format = '.webp'
  }

  const getAssets = (imageDir) => {
    return fs.readdirSync(imageDir) // Read the contents of the assets folder
  }

  const getPicture = async (img, src) => {

    // If the image is referencing an external url, skip it
    if (img.attrs.src.startsWith('http://') || img.attrs.src.startsWith('https://')) {
      return img
    }

    // If the image has a skip flag, skip it
    if(img.attrs['skip'] == true) {
      return img
    }

    // Loop through the different set sizes and check if an image exists for that size
    let srcset = []
    var dimensions = sizeOf(src)
    var relativeSrc = src.split('src')[1]
    img.content = [];

    if (src.indexOf('.gif') != -1) {
      img['attrs']['src'] = relativeSrc
      return img
    }
    

    if(typeof(img['attrs']['fullsize']) == "undefined"){
      // Loop through the assets folder and see if each size exists in webp format
      for (var i = 0; i < imageOptions.size.length; i++) {
        var size = imageOptions.size[i]
        var suffix =  '@' + size + 'w';
        var image =
          relativeSrc.replace(/\.(\w+)$/, suffix + options.format) +
          ' ' +
          size +
          'w'

        if(!(size > dimensions.width)) {
          srcset.push(image)
        }
      }
    }

    // add full size to images array
    srcset.push(relativeSrc.replace(/\.(\w+)$/, options.format+' '+dimensions.width+'w'));

    // Generate the picture tag with the <source> elements + the <img> element
    img.tag = 'picture'
    delete img['attrs']['src'];

    // Add an alt tag if it's left empty
    if(img['attrs']['alt'] === undefined) {
      img['attrs']['alt'] = ''
    }


    if(img['attrs']['nolazy'] !== undefined) {
      img['attrs']['src'] = relativeSrc.replace(/\.(\w+)$/, options.format)
    } else {
      if(typeof(img['attrs']['class']) == 'undefined'){
        img['attrs']['class'] = '';
      }

      img['attrs']['class'] += ' lazyload';
    }

    // Set the srcset <source> tags
    let srcsetTag = {
      tag: 'source',
      attrs: {
        "data-srcset": srcset.join(', '),
        type: 'image/webp'
      },
    };


    // Set the fallback image

    let fallbackType = relativeSrc.split('.')[1];
    if(fallbackType == 'jpg'){
      fallbackType = 'jpeg';
    }

    let defaultSrc =  {
      tag: 'source',
      attrs: {
        "data-srcset": relativeSrc,
        type: 'image/'+ fallbackType
      },
    };

    if(img['attrs']['nolazy'] !== undefined) {
      defaultSrc['attrs']['srcset'] = defaultSrc['attrs']['data-srcset'];
      srcsetTag['attrs']['srcset'] = srcsetTag['attrs']['data-srcset'];
    }

    img.content.push([
      srcsetTag,
      defaultSrc,
    ]);
  
    img.content.push([
      {
        tag: 'img',
        attrs: {
          "data-src": relativeSrc.replace(/\.(\w+)$/, '.webp'),
          skip: true,
          width: dimensions.width,
          height: dimensions.height,
          ...img['attrs']
        },
      },
    ]);

    if (imageOptions.generate == false) {
      img['attrs']['src'] = relativeSrc
      img.content = [
        {
          tag: 'img',
          attrs: {
            "data-src": relativeSrc,
            skip: true,
            width: dimensions.width,
            height: dimensions.height,
           ...img['attrs']
          },
        },
      ]
    }

    delete img['attrs'];
    return img
  }

  /**
   * Ease of use function to generate all the images + sizes
   */
  const generateImages = async () => {
    if (imageOptions.generate) {
      return await generateAll(imageOptions)
    }
  }

  return (tree) =>
    new Promise((resolve, reject) => {
      // generates the required image files
      generateImages()

      const promises = []
      if (!tree.promises) tree.parser = parser
      if (!tree.match) tree.match = match

      tree.match({ tag: 'img' }, (node) => {
        if (!node.attrs['skip'] && !node.attrs.src?.includes('.svg')) {
          promises.push(
            new Promise(async (resolve, reject) => {
              try {
                var src = node.attrs['src']
                if (src) {
                  var imgSrc = path.resolve(cwd, src)
                  // Get the picture tag and assign it
                  node = await getPicture(node, imgSrc)
                }
                resolve()
              } catch (err) {
                reject(err)
              }
            })
          )
        }
        return node
      })
      Promise.all(promises)
        .then(() => resolve(tree))
        .catch(reject)
    })
}
